runtime/domain.c

```diff
@@ -45,20 +45,26 @@
 #include "caml/sync.h"
 #include "caml/weak.h"
 
-/* From a runtime perspective, domains must handle stop-the-world (STW)
-   sections, during which:
-    - they are within a section no mutator code is running
-    - all domains will execute the section in parallel
-    - barriers are provided to know all domains have reached the
-      same stage within a section
+/* The runtime can run stop-the-world (STW) sections, during which all
+   active domains run the same code in parallel (with a barrier
+   mechanism to synchronize within this code). See
+   [caml_try_run_on_all_domains] below.
 
    Stop-the-world sections are used to handle duties such as:
     - minor GC
-    - major GC to trigger major state machine phase changes
-
-   Two invariants for STW sections:
-    - domains only execute mutator code if in the stop-the-world set
-    - domains in the stop-the-world set guarantee to service the sections
+    - major GC phase changes
+
+   We guarantee that no mutator code runs in parallel with a STW
+   section, and domains are blocked from entering or leaving the set of active
+   STW participants while a STW section is in progress.
+
+   To provide these guarantees:
+    - domains must register as STW participants before running any
+      mutator code
+    - domains registered as STW participants must be careful to
+      service STW interrupt requests
+    - STW sections must not trigger callbacks (eg. finalisers or
+      signal handlers).
```

Comment:
Would it be possible to explicitly mention an example of a domain that is "running" but is not part of the STW participants? A newly created domain that is yet to enter OCaml I guess. Since this domain is not part of the STW participants, what is this domain doing when the STW section is ongoing? Is it parked on a particular condition variable?

Reply:
These more detailed aspects are discussed in the documentation comment of `caml_try_run_on_all_domains_with_spin_work`. Rather than try to summarize them in the overview comment, I made the forward reference to that detailed documentation more explicit.

>    See the comments on [caml_try_run_on_all_domains_with_spin_work]
   below for more details on the synchronization mechanisms involved.